internal/pkg/keystore/pkcs12/pkcs12.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2021 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pkcs12
+
+import (
+	"crypto"
+	"crypto/rand"
+	"errors"
+	"fmt"
+
+	"github.com/cert-manager/cert-manager/pkg/util/pki"
+	"software.sslmate.com/src/go-pkcs12"
+)
+
+// Create combines the inputs to a single PKCS12 keystore file.
+// key must be PKCS1 or PKCS8 encoded. certificates must be PEM encoded.
+func Create(key crypto.PrivateKey, chainPEM []byte, rootPEM []byte) ([]byte, error) {
+	if key == nil {
+		return nil, errors.New("key must not be nil")
+	}
+
+	if len(chainPEM) == 0 {
+		return nil, errors.New("chainPEM must not be empty")
+	}
+
+	if len(rootPEM) == 0 {
+		return nil, errors.New("rootPEM must not be empty")
+	}
+
+	root, err := pki.DecodeX509CertificateBytes(rootPEM)
+	if err != nil {
+		return nil, fmt.Errorf("pki.DecodeX509CertificateChainBytes(rootPEM): %v", err)
+	}
+
+	chain, err := pki.DecodeX509CertificateChainBytes(chainPEM)
+	if err != nil {
+		return nil, fmt.Errorf("pki.DecodeX509CertificateChainBytes(chainPEM): %v", err)
+	}
+
+	leaf := chain[0]
+	chain = chain[1:]
+
+	// add the root cert to the back of the chain
+	chain = append(chain, root)
+
+	pfx, err := pkcs12.Encode(rand.Reader, key, leaf, chain, pkcs12.DefaultPassword)
```

Comment:
I think we want the leaf + chain and then the root, rather than the leaf and then the root + chain (or if we can't do that, to omit the chain).

What do we do in cert-manager's Certificate issuing controller currently? 👀

Reply:
That is not how the lib works when encoding: https://pkg.go.dev/software.sslmate.com/src/go-pkcs12#Encode

Reply:
I wonder if we should not include the certificate chain then - clients should really be providing the chain along with the leaf certificate when performing clientAuth, or as a server should present the leaf+chain to the client. The chain is by definition intermediate certificates, not roots.

Reply:
The csi-driver is already writing the key, chain and CA. Why would we not write them all for p12?

Reply:
I think the proposed change should go here instead: https://github.com/cert-manager/csi-driver/blob/bfbdf99951e935ffde2bb3c1524a4a9ae295941c/pkg/filestore/writer.go#L89

i.e. let the user (a programmer) choose to include the root or not.

@JoshVanL Do you know if excluding the root is the right thing to do?

Reply:
Thanks for the elaboration!